Adding cli command for UI and handling user input with factstore usin…

zmhassan opened the pull request:
# PR Details

For our system to scale we should look at better options then filebacked storage. With this PR I'll be providing a database backed storage for the fact_store.

## Description

We are using SQLAlchemy which supports different databases its widely used orm. You just need to update 1 environment variable and you can connect to many popular databases.

## Related Issue
Fixes #83 

## Motivation and Context

When we are running this system at scale we ran into issues with elasticsearch. If we push to many messages it fails. Also there is a limit on how many messages we can pull down. In that case we decided to skip elasticsearch and track our data in fact_store. Databases are a good location to store your data.


## How Has This Been Tested

This feature includes many moving parts and you'll need to setup your favorite database. In my case I'm using MySQL. In that case you can try the following to test this on mysql.
```bash

export SQL_CONNECT="mysql+pymysql://{USERNAME}:{PASSWORD}@localhost/factstore"
pip install --editable .
scorpio ui --port 8080 --debug True

```

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.


## More Details

Try this out. Let me know what you think.

vpavlin commented:
Please fix linting before requesting review - this is a bit confusing to read:)

Also, once again, try to stick to 1 change per PR - I see some reformatting of HTML and JavaScript and adding try/except which ends up in reformating most of the file as well - was that necessary for the database addition? 

As the primary deployment target is OpenShift - how do you plan on deploying the database and connecting the services? I don't see any documentation for that or a deployment script or something like that

zmhassan commented:
Hi @vpavlin 

Thank you for the quick feedback. As I explained in the title [WIP]. I pinged you'll to take a first pass through this as a way for us to work better/faster together. 

I'll explain each change (ignoring the code formatting as its good to do):

1. **UI Code:**
I added a new field to the UI so when users of the service say this event is a false anomaly they can also include notes so we can look at later of why it's a false anomaly. 
2. **FactStoreDB:**
I can include an additional file with documentation on how to configure the fact store with a database. So since we use sqlachemy, users can pick their favourite database and just provide a connection string and that should be easy.
3. **CLI:**
I added a command to the cli which lets users testing out this service on their laptop to be able to run the fact_store ui via command 'scorpio ui ' in one terminal and 'scorpio run --job-type train' in another window.

So the goal I have for this system is to make both the enduser benefit from being able to tag their anomaly as false or positive to get more accurate notifications. And for our data scientists who will join us to have the tooling to look back at the data what is not working in terms of false predictions.

zmhassan commented:
@vpavlin  
Thank you for taking the time to review this PR.

- [ ]  I've included all fixes to code linting
- [ ] I've added some documentation on how to run factstore locally with database.
- [ ] I've included more tests to validate that metadata persists to factstore.
- [ ] I've validated that bloom filter checks if we have duplicates predict_id's being persisted and throw exception. This saves us time in repeat db requests.

zmhassan commented:
FYI if you would like to test the fact_store and deploy on openshift run:

```bash
oc new-app python~https://github.com/zmhassan/log-anomaly-detector.git#issue_83  -e APP_FILE='fact_store/app.py' --name='factstore-zmhassan'
```

vpavlin commented on fact_store/model.py:
```diff
@@ -0,0 +1,56 @@
+                primary_key=True,
+                autoincrement=True)
+    # Required
+    predict_id = Column(String(255),
```
Should the `predict_id` be an FK for Event `predict_id`?

zmhassan replied:
Yes it should.

zmhassan replied:
Yes it should be a foriegn key.

zmhassan replied:
Fixed

vpavlin commented on fact_store/model.py:
```diff
@@ -0,0 +1,56 @@
+    We track this for quality control
+    """
+    __tablename__ = "events"
+    id = Column(Integer, primary_key=True, autoincrement=True)
```
Why use separate `id` if `predict_id` is unique anyway?

zmhassan replied:
We can make the predict_id unique and yes we can make it the primary key

zmhassan replied:
Fixed.

vpavlin commented on fact_store/fact_store_api.py:
```diff
@@ -0,0 +1,85 @@
+    anomaly-detection
+
+    """
+    f = BloomFilter(capacity=1000, error_rate=0.001)
```
What happens on the app restart? We start from empty BF? In that case `is_not_anomaly` does not really work, does it?

zmhassan replied:
Good catch that is one of the edge cases. at startup I can load the bloom filter with all the feedback predict_ids. Another thing is I don't expect the feedback to be more then 1000.

zmhassan replied:
@vpavlin I've opened an issue for this one. #109

https://github.com/AICoE/log-anomaly-detector/issues/109

I'll work on this as a separate PR. I want to keep this PR to just what the title describes and the git issue described associated with this PR.

vpavlin commented on tests/test_factstore.py:
```diff
@@ -1,7 +1,62 @@
 
+def test_feedback_inserted():
+    fs = FactStore(True)
+    id1= uuid.uuid4()
```
missing space

zmhassan replied:
Fixed

vpavlin commented on tests/test_factstore.py:
```diff
@@ -1,7 +1,62 @@
-from fact_store.app import FactStore
+from fact_store.fact_store_api import FactStore
```
I am not sure how useful these tests are - you are basically just trying out the SQLAlchemy and making sure it creates the tables as defined and can perform an insert and select - i.e. testing SQLAlchemy itself rather than your own code

zmhassan replied:
The test cases are there to validate that when I want to store feedback that it persists. Also checking to make sure that we don't get duplicate feedback. Please give the code another read.

zmhassan replied:
I've reduced the amount of repeat tests and will keep it simple for now. Is this a blocker?

zmhassan commented on app.py:
```diff
@@ -12,6 +13,19 @@ def cli():
+@click.option("--port", default=5001,
+              help="Select the port number you would like to run the web ui ")
+
+def ui( debug, port):
```
Good catch I'll fix this

zmhassan commented on fact_store/README.md:
```diff
@@ -2,12 +2,45 @@
+By default if don't provide a database we get sqlite as our default storage
+If you would like to utilize a different storage database that is supported 
+by sqlalchemy. 
+
```
Fixing this now

zmhassan replied:
Done

zmhassan commented on fact_store/README.md:
```diff
@@ -2,12 +2,45 @@
+
+export SQL_CONNECT="mysql+pymysql://{USERNAME}:{PASSWORD}@localhost/factstore"
+pip install --editable .
+scorpio ui --port 8080 --debug True
```
Its the binary cli that gets generated. I'll update this to `python app.py ui`

zmhassan replied:
Done !

zmhassan commented on fact_store/app.py:
```diff
@@ -107,35 +21,56 @@ def metadata():
+    return jsonify({"feedback":df})
 
 
 @app.route("/api/feedback", methods=['GET'])
```
Yea. We can do that.

zmhassan replied:
I fixed the issue. Now its a post

zmhassan commented on fact_store/app.py:
```diff
@@ -107,35 +21,56 @@ def metadata():
+        if not id or not anomaly:
+            raise Exception('This service requires that you provide the id' +
+                            ', anomaly=True|False and notes are optional ')
+        print("id: {} ".format(id))
```
Good catch. I'll fix that

zmhassan commented:
@vpavlin @durandom I've fixed all issues raised lets press forward on this.

durandom commented on fact_store/model.py:
```diff
@@ -0,0 +1,57 @@
+        return {c.name: getattr(self, c.name) for c in self.__table__.columns}
+
+
+
```
empty lines at end of file

zmhassan replied:
Fixed

durandom commented on tests/test_factstore.py:
```diff
@@ -1,57 +1,51 @@
-from fact_store.app import FactStore
+import uuid
```
seems not to be used, remove

zmhassan replied:
Fixed.

zmhassan commented on fact_store/app.py:
```diff
@@ -1,96 +1,11 @@
+from fact_store.fact_store_api import FactStore
+app = Flask(__name__)
+
 app = Flask(__name__, static_folder="static")
```
Fixed.

zmhassan commented on fact_store/app.py:
```diff
@@ -107,38 +22,60 @@ def metadata():
+        content = request.json
+        # id = request.args.get('lad_id')
+        # anomaly = bool(request.args.get('is_anomaly'))
+        # notes = request.args.get('notes')
```
Fixed.

zmhassan commented on fact_store/fact_store_api.py:
```diff
@@ -0,0 +1,85 @@
+        event = EventModel(message=message,
+                               score=score,
+                               predict_id=predict_id,
+                               anomaly_status=anomaly_status)
```
Fixed

zmhassan commented on fact_store/fact_store_api.py:
```diff
@@ -0,0 +1,85 @@
+        return _id in self.f
+
+
+
```
Fixed

durandom replied:
actually no. there are still 3 empty lines :(

zmhassan commented:
> just some small style nitpicks.
> please fix them in a _new_ commit, so that we see what has been changed :)

No problem. I've fixed issues and have made changes in the 2nd commit. Hope you can review again and we can close this for good.

durandom commented on tests/test_factstore.py:
```diff
@@ -10,23 +10,23 @@ def test_events_inserted():
-    # TODO: Delete all rows currently in the store.
-    fs.session.query(EventModel).delete()
-    fs.write_event(predict_id="a2b35c5b-016d-4e2c-8ec5-87d1b962b2f8",
+    fact_store = FactStore(True)
```
+1 for using better variable_names
